dot/sync/chain_sync.go

```diff
@@ -843,17 +843,41 @@ func (cs *chainSync) processBlockData(blockData types.BlockData, origin blockOri
 	announceImportedBlock := cs.getSyncMode() == tip
 
 	if blockData.Header != nil {
+		var setFinalisedHash func() error = nil
+		if blockData.Justification != nil && len(*blockData.Justification) > 0 {
+			round, setID, err := cs.finalityGadget.VerifyBlockJustification(
+				blockData.Header.Hash(), blockData.Header.Number, *blockData.Justification)
+			if err != nil {
+				return fmt.Errorf("verifying justification: %w", err)
+			}
+
+			setFinalisedHash = func() error {
+				header := blockData.Header
+
+				err = cs.blockState.SetFinalisedHash(header.Hash(), round, setID)
+				if err != nil {
+					return fmt.Errorf("setting finalised hash: %w", err)
+				}
+
+				err = cs.blockState.SetJustification(header.Hash(), *blockData.Justification)
+				if err != nil {
+					return fmt.Errorf("setting justification for block number %d: %w", header.Number, err)
+				}
+
+				return nil
+			}
+		}
+
 		if blockData.Body != nil {
 			err := cs.processBlockDataWithHeaderAndBody(blockData, origin, announceImportedBlock)
 			if err != nil {
 				return fmt.Errorf("processing block data with header and body: %w", err)
 			}
 		}
 
-		if blockData.Justification != nil && len(*blockData.Justification) > 0 {
-			err := cs.handleJustification(blockData.Header, *blockData.Justification)
-			if err != nil {
-				return fmt.Errorf("handling justification: %w", err)
+		if setFinalisedHash != nil {
```

Comment:
Why we cannot just move the finalization logic here instead of using the closure?

Reply:
@EclesioMeloJunior thoughts?

Reply:
not a move but will looks like this:

```
round, setID uint64
if hasJustification() {
   round, setID := validate()
}

if hasBody() {
    import()
}

if hasJustification() {
   setFinalizedHash(round, setID)
}
```

Reply:
How is substrate handling this logic?

Reply:
I made some changes in https://github.com/ChainSafe/gossamer/pull/4144/commits/104abb478be4302906aa6e9f386da31af1f162a9

Reply:
I don’t think the code is hard to get it. As I’ve mentioned in the pseudo code in the previous comment, the closure captures the variables in a context that can be used later. Looking into how substrate manage this logic is not worth given the differences in both implementations, but I am pretty sure the flow is the same, validate -> import -> persist finalization